File: sdk/js/examples/03_node_client_api/students.ts
// Copyright 2021 Touca, Inc. Subject to Apache-2.0 License.

export interface Student {
  username: string;
  fullname: string;
  dob: Record<'year' | 'month' | 'day', number>;
  height: number;
  weight: number;
  courses: {name: string, grade: number}[]
}

const students: Student[] = [
  {
    username: 'alice',
    fullname: 'Alice Anderson',
    dob: { year: 1980, month: 3, day: 1 },
    height: 6.3,
    weight: 152,
    courses: [{name: 'math', grade: 4.0}]
  },
  {
    username: 'bob',
    fullname: 'Bob Brown',
    dob: { year: 1980, month: 6, day: 31 },
    height: 6.0,
    weight: 145,
    courses: [{name: 'english', grade: 3.7}, {name: 'history', grade: 3.9}]
  },
  {
    username: 'charlie',
    fullname: 'Charlie Clark',
    dob: { year: 1979, month: 9, day: 19 },
    height: 5.5,
    weight: 120,
    courses: [{name: 'math', grade: 2.9}, {name: 'computers', grade:3.7}]
  }
];

export function parse_profile(username: string): Student | undefined {
  return students.find((item) => item.username === username);
}

File: examples/python/02_python_basic_api/students.py
# Copyright 2021 Touca, Inc. Subject to Apache-2.0 License.

from dataclasses import dataclass
from time import sleep
from typing import List
import touca


@dataclass
class Date:
    year: int
    month: int
    day: int


@dataclass
class Student:
    username: str
    fullname: str
    dob: Date
    gpa: float


@dataclass
class Course:
    name: str
    grade: float


@dataclass
class StudentData:
    username: str
    fullname: str
    dob: Date
    courses: List[Course]


students = [
    StudentData(
        "alice",
        "Alice Anderson",
        Date(2006, 3, 1),
        [Course("math", 4.0), Course("computers", 3.8)],
    ),
    StudentData(
        "bob",
        "Bob Brown",
        Date(1996, 6, 31),
        [Course("english", 3.7), Course("history", 3.9)],
    ),
    StudentData(
        "charlie",
        "Charlie Clark",
        Date(2003, 9, 19),
        [Course("math", 2.9), Course("computers", 3.7)],
    ),
]


def calculate_gpa(courses: List[Course]):
    touca.add_result("courses", courses)
    return sum(k.grade for k in courses) / len(courses) if courses else 0


def parse_profile(username: str) -> Student:
    sleep(0.1)
    data = next((k for k in students if k.username == username), None)
    if not data:
        raise ValueError(f"no student found for username: ${username}")
    return Student(data.username, data.fullname, data.dob, calculate_gpa(data.courses))
